### ISSUE TITLE
Consolidate UserStore content across landscape, preparing to join

### ISSUE THREAD

**Issue by eclipse-sailing-analytics-bot** [eclipse-sailing-analytics/sailing-analytics#4006]
| Field | Value |
|---|---|
| **Reporter** | @fmittag |
| **Assignee** | @axeluhl |
| **Product** | Sailing Race Analytics |
| **Component** | User and Account Management |
| **Version** | unspecified |
| **Hardware** | All |
| **OS** | All |
| **Severity** | enhancement |
| **Priority** | P1 |
| **CC (unmapped)** | Steffen Schäfer (`steffen.schaefer@pgt.de`) |
| **Created** | 2016-12-19T13:52:24Z |
| **Original Bug** | [Bug 4006](https://old-bugzilla.sapsailing.com/bugzilla/show_bug.cgi?id=4006) |

---

In 2016 we introduced public user accounts to enable users to register at sapsailing.com. We have seen at least some registrations but all these registrations happened on our live servers. As we have no central user management yet these users get lost when the events are transferred from the live servers to our archive server. This means that a registered user can't access his account anymore. Therefore we should at least transfer (export/import) these users in some way from all the live server DB's to the archive server DB.
Furthermore we need to make sure that we don't overwrite admin accounts on the archive server.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

Commits was 438d69e689c92b2a0f92fb55bbea1700642696e6.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

Added tests validating that:
- ACL on dropped User / UserGroup is dropped
- Ownership on dropped User / UserGroup is dropped
- Ownerships on role assignment and permission assignment of a dropped user are dropped

Other tests that have been added:
- no role association ownership or ACL ends up in target for a role dropped because it lost its qualifications or was unqualified in the first place
- no permission association ownership or ACL ends up in target for a permission dropped because it had no object constraint

With commit 7f86a28cbf260f2ec10db12221f6b68032dc80b4 all tests are green.

Next steps:
 - create a copy of all DBs in question
 - run the merge on the copies, producing a candidate for the central security service
 - Compare the outcome to expectations: anything broken? Can a server be launched off the resulting DB? Manual spot checks!
 - Launch a server as a replica for SecurityService and see if things work

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

The replication service waits for the Replicable objects to appear in the OSGi service registry. With this, other service trackers will then also resolve them and let other services call methods on them.

Both, the writing and the reading method calls need to be deferred until the replicable has been fully initialized.

What if the Replicable was only the data holder, and the actual service interface was a separate OSGi service that made itself available in the registry only after any requested replication of its underlying data service has completed?

Alternatively, we could introduce smart Replicable service trackers which do two things:
 1) using an OSGi service tracker to wait for and obtain a reference to the Replicable service as it appears in the OSGi registry
 2) if the Replicable has auto-replicate configured, return the reference to the Replicable only after the replication has been fully initialized

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

Looking at the two cases "partial" and "full" replication, where in both cases the SecurityService and SharedSailingData will be replicated, but RacingEventService will only be replicated with "full" replication:

 - partial: RacingEventServiceImpl obtains the SecurityService that was initialized from the local DB; after loading the local RacingEventServiceImpl content from the local DB, ensureOwnership(...) is called. But worse, for each restored race, after the TrackedRace has been created, a call to getSecurityService().migrateOwnership(trackedRace) occurs. The SecurityService obtained this way could be in any state, such as still based on the local DB, or empty because it was cleared before initial load, or in an inconsistent state while clearing the state loaded locally and while replacing it from the initial load received.
 
  - full: the loading of the RacingEventServiceImpl from local contents could still load content that the "phony" local replica DB contains. Ownership migrations may be attempted, and at that point in time it is unclear what about the replication status of the SecurityService. If migrating ownerships against an empty or locally-initialized SecurityService where the true ownerships live in the central shared SecurityService would lead to ownership migrations being attempted against a SecurityService in unclear and possibly changing state.
  
In both cases, access to a Replicable that is expecting the replicable to have consistent content and be able to accept "transactions" has to wait for the Replicable to have finished its replication business. Such replicables need to become unavailable when it is clear that replication is about to start (before clearing their state even), and have to become available not before the initial load has completed.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

Commit 39608e3f6cf481fcc4541c67a6e5dde5bcd65203 on [bug4006](https://github.com/eclipse-sailing-analytics/sailing-analytics/tree/bug4006) suggests an implementation based on a specialization of ServiceTracker which for Replicables can wait for the replicable *and* for the ReplicationService to not be in startup mode. The RacingEventService has been modified to use this specialized tracker and to wait for fully-initialized SecurityService and SharedSailingData instances. Let's see what the tests look like, locally as well as for https://hudson.sapsailing.com/job/[bug4006](https://github.com/eclipse-sailing-analytics/sailing-analytics/tree/bug4006)/51/.

**Comment by eclipse-sailing-analytics-bot** [comment]
**@axeluhl** commented:

In commits up to a7b8262a8b5c508e9e7f20582e89a2c84f4f101c I converted all ServiceTracker constructions for Replicable objects into FullyInitializedReplicableTracker usages.
